### ISSUE TITLE
Refactor 1D solvers

### ISSUE THREAD

**Issue by djsilva99** [djsilva99/heatrapy#45]
# Description
Heatrapy needs to be cleaned to make the exception's logic more sound, and to increase the test coverage. This issue addresses the refactoring of 1D solvers.

**Comment by formeo** [comment]
Hi Daniel, I'd like to help with this. I profiled the 1D solvers and found
two main issues:
 
1. **Duplicated latent heat block** — the same ~27 lines of phase-transition
   logic are copy-pasted in all 4 solvers. Any bug fix has to be applied
   4 times.
 
2. **`copy.deepcopy` + Python loops in explicit solvers** — each call to
   `explicit_general` / `explicit_k` does a deepcopy of the temperature array,
   then 3 separate Python-level loops for computing, copying, and repacking.
 
My plan:
 
- Extract the latent heat block into a shared `_latent_heat.py` helper
- Vectorize the explicit FDM stencils with numpy (the implicit solvers
  already use numpy for the tridiagonal solve, so they just get the
  deduplication)
- Keep the same function signatures and return format, no API changes
 
All 25 existing tests pass with this change. I also saw a minor 1.6×
speedup on the test suite (118 → 72s), though that'll be more significant
on larger grids.
 
Would you be happy with a PR along these lines?

**Comment by djsilva99** [comment]
Hi @formeo 
Sure, feel free to create the Pull Request.
In adition, I would like to handle the typing of the arguments, which is a mess right now, but we can leave it to another PR.

**Comment by formeo** [comment]
Thanks Daniel! PR is up. And yeah, the typing cleanup makes sense as a separate pass — happy to help with that too once this one lands.

**Comment by formeo** [comment]
Since #73 is merged, I'd like to tackle the typing part of this issue.

I've looked through the codebase — the main areas without type hints are:
- `dimension_1/solvers/` (4 solver functions + `_latent_heat.py`)
- `dimension_1/objects/` (SingleObject, SystemObject)
- `mats/calmatpro.py`

I'll start with the solvers module (since I'm already familiar with it from #73) and work outward. One PR covering all modules, or would you prefer separate PRs per directory?

**Comment by djsilva99** [comment]
It seems better to split this into separate PRs, so please create one PR just for the solvers here, since we’re addressing this solvers issue. Thanks for the contribution! Don’t forget to raise a TypeError, not a ValueError, when a variable has the wrong type.

**Comment by formeo** [comment]
Solvers PR is up: #75

Covers all 5 files in `dimension_1/solvers/`, including TypeError checks
and 14 new unit tests. Will continue with `objects/` and `mats/` in
separate PRs after this one lands.
